File: schema_salad/java/main_utils/ArrayLoader.java
package ${package}.utils;

import java.util.ArrayList;
import java.util.List;

public class ArrayLoader<T> implements Loader<List<T>> {
  private final Loader<T> itemLoader;
  private final boolean flatten;

  public ArrayLoader(Loader<T> itemLoader) {
    this(itemLoader, true);
  }

  public ArrayLoader(Loader<T> itemLoader, boolean flatten) {
    this.itemLoader = itemLoader;
    this.flatten = flatten;
  }

  public List<T> load(
      final Object doc,
      final String baseUri,
      final LoadingOptions loadingOptions,
      final String docRoot) {
    final List<Object> docList = (List<Object>) Loader.validateOfJavaType(List.class, doc);
    final List<T> r = new ArrayList();
    final List<Loader> loaders = new ArrayList<Loader>();
    loaders.add(this);
    loaders.add(this.itemLoader);
    final UnionLoader unionLoader = new UnionLoader(loaders);
    final List<ValidationException> errors = new ArrayList();
    for (final Object el : docList) {
      try {
        final Object loadedField = unionLoader.loadField(el, baseUri, loadingOptions);
        if (this.flatten && loadedField instanceof List) {
          r.addAll((List<T>) loadedField);
        } else {
          r.add((T) loadedField);
        }
      } catch (final ValidationException e) {
        errors.add(e);
      }
    }
    if (!errors.isEmpty()) {
      throw new ValidationException("", errors);
    }
    return r;
  }
}

File: schema_salad/typescript/util/loaders/ArrayLoader.ts
import { Loader, loadField, LoadingOptions, _UnionLoader, ValidationException } from '../Internal'

export class _ArrayLoader implements Loader {
  items: Loader[]
  flatten: boolean

  constructor (items: Loader[], flatten: boolean = true) {
    this.items = items
    this.flatten = flatten
  }

  async load (doc: any, baseuri: string, loadingOptions: LoadingOptions, docRoot?: string): Promise<any> {
    if (!Array.isArray(doc)) {
      throw new ValidationException('Expected a list')
    }
    let r: any[] = []
    const errors: ValidationException[] = []
    for (var val of doc) {
      try {
        const lf = await loadField(val, new _UnionLoader([this, ...this.items]), baseuri, loadingOptions)
        if (this.flatten && Array.isArray(lf)) {
          r = r.concat(lf)
        } else {
          r.push(lf)
        }
      } catch (e) {
        if (e instanceof ValidationException) {
          errors.push(e)
        } else {
          throw e
        }
      }
    }
    
    if (errors.length > 0) {
      throw new ValidationException('', errors)
    }
    return r
  }
}
